Rewrite RhUnbox in managed code

MichalStrehovsky opened the pull request:
Mostly want to get rid of the dependencies on nullable-related fields on EEType on the unmanaged side of the runtime. This is the only place that depends on them. Deletion will follow.

MichalStrehovsky commented on src/Runtime.Base/src/System/Runtime/InternalCalls.cs:
```diff
@@ -190,6 +191,17 @@ internal static int RhEndNoGCRegion()
+        [RuntimeImport(Redhawk.BaseName, "RhpInitMultibyte")]
+        internal static extern unsafe void RhpInitMultibyte(ref byte dmem, int c, nuint size);
+
+        [DllImport(Redhawk.BaseName)]
```
Would it be okay to make this RuntimeImport?

jkotas replied:
I think so. 

RuntimeImport is a bit problematic for GC suspension, but we have same shortcut in many other places. The right solution is to make RuntimeImport GC suspension friendly by teaching JIT how to probe for GC around it efficiently.

jkotas replied:
This is clone of the memmove from the CoreLib. The one in CoreLib needs to stay as `DllImport`. Skipping GC probing around that one would be too impactful.

jkotas commented on src/Runtime.Base/src/System/Runtime/RuntimeExports.cs:
```diff
@@ -216,7 +222,65 @@ public static unsafe void RhUnboxNullable(ref byte data, EETypePtr pUnboxToEETyp
+
+                // The first field of the Nullable is a Boolean which we must set to false in this case to indicate no
+                // value is present.
+                Unsafe.As<byte, bool>(ref data) = false;
```
You can simplify this by zeroing out the whole thing. Zeroing out a few more bytes is going to be smaller and faster than doing the math with NullableType, NullableValueOffset, etc.

MichalStrehovsky replied:
Yeah, the code was trying to be too explicit. Done.

MichalStrehovsky commented:
@yowl, could you have a look at e2c3326?

I started using `memmove` in another location where the existing workaround didn't work out so well.

This made it work in the frankenbuild that we test right now "Windows x64 CoreLib compiled into 32bit WASM", but I'm not sure whether this could break the real WASM.

```
2020-03-14T19:31:25.8874328Z   CpObj -> F:\workspace\_work\1\s\tests\src\Simple\HelloWasm\bin\Debug\x64\CpObj.dll
2020-03-14T19:31:25.9170027Z   ILHelpers -> F:\workspace\_work\1\s\tests\src\Simple\HelloWasm\bin\Debug\x64\ILHelpers.dll
2020-03-14T19:31:26.5698356Z   Generating native code
2020-03-14T19:31:27.3272213Z   The method or operation is not implemented. ([S.P.CoreLib]System.Runtime.RuntimeExports.RhUnbox(object,uint8&,EEType*))
2020-03-14T19:31:27.4017436Z   Unable to cast object of type 'Internal.IL.Stubs.StartupCode.StartupCodeMainMethod' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([HelloWasm]<Module>.StartupCodeMain(int32,native int))
2020-03-14T19:31:34.0148204Z   The method or operation is not implemented. ([S.P.CoreLib]System.Runtime.RuntimeImports.memmove(uint8*,uint8*,uint64))
2020-03-14T19:31:38.2901270Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseDelegateStub__EnumCalendarInfoProcExEx(char*,uint32,native int,void*))
2020-03-14T19:31:38.2903658Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseOpenStaticDelegateStub__EnumCalendarInfoProcExEx(char*,uint32,native int,void*))
2020-03-14T19:31:44.9311567Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseDelegateStub__EnumTimeFormatsProcEx(char*,void*))
2020-03-14T19:31:44.9315794Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseOpenStaticDelegateStub__EnumTimeFormatsProcEx(char*,void*))
2020-03-14T19:31:53.3580068Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseDelegateStub__EtwEnableCallback(Guid*,int32,uint8,int64,int64,EVENT_FILTER_DESCRIPTOR*,void*))
2020-03-14T19:31:53.3602458Z   Unable to cast object of type 'Internal.IL.Stubs.DelegateMarshallingMethodThunk' to type 'Internal.TypeSystem.Ecma.EcmaMethod'. ([S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseOpenStaticDelegateStub__EtwEnableCallback(Guid*,int32,uint8,int64,int64,EVENT_FILTER_DESCRIPTOR*,void*))
2020-03-14T19:32:16.3575936Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3580547Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3583712Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3585169Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 273, scope: !483177)
2020-03-14T19:32:16.3586372Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3587071Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3587638Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3589479Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 271, scope: !483177)
2020-03-14T19:32:16.3590329Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3590918Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3591434Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3592125Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 270, scope: !483177)
2020-03-14T19:32:16.3592876Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3593404Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3594202Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3595093Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 268, scope: !483177)
2020-03-14T19:32:16.3595853Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3596345Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3596793Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3597504Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 262, scope: !483177)
2020-03-14T19:32:16.3598299Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3598810Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3599274Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3600044Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 255, scope: !483177)
2020-03-14T19:32:16.3600771Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3601234Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3601633Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3602182Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 254, scope: !483177)
2020-03-14T19:32:16.3602776Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3603219Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3603589Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3604173Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 252, scope: !483177)
2020-03-14T19:32:16.3604856Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3605352Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3605771Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3606417Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 248, scope: !483177)
2020-03-14T19:32:16.3607141Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3607618Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3608014Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3608741Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 237, scope: !483177)
2020-03-14T19:32:16.3609554Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3610212Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3610916Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3611640Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 234, scope: !483177)
2020-03-14T19:32:16.3612369Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3612838Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3613274Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3613911Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 233, scope: !483177)
2020-03-14T19:32:16.3614592Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3615058Z   void ()* @llvm.donothing
2020-03-14T19:32:16.3615488Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3616171Z     call addrspace(0) void @llvm.donothing(), !dbg !DILocation(line: 230, scope: !483177)
2020-03-14T19:32:16.3616914Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3617529Z   %"[S.P.CoreLib]Internal.Runtime.EEType"* (i8*)* @Object__get_EEType
2020-03-14T19:32:16.3618074Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3618979Z     <badref> = call addrspace(0) %"[S.P.CoreLib]Internal.Runtime.EEType"* @Object__get_EEType(i8* <badref>), !dbg !DILocation(line: 245, scope: !483177)
2020-03-14T19:32:16.3619865Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3620606Z   i1 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_IsValueType
2020-03-14T19:32:16.3621126Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3621975Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__get_IsValueType(i8* <badref>), !dbg !DILocation(line: 248, scope: !483177)
2020-03-14T19:32:16.3622907Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3623455Z   i1 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable
2020-03-14T19:32:16.3623977Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3624796Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable(i8* <badref>), !dbg !DILocation(line: 253, scope: !483177)
2020-03-14T19:32:16.3625657Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3626186Z   i1 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable
2020-03-14T19:32:16.3626697Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3627486Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable(i8* <badref>), !dbg !DILocation(line: 252, scope: !483177)
2020-03-14T19:32:16.3628350Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3628873Z   i1 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable
2020-03-14T19:32:16.3629378Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3630165Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__get_IsNullable(i8* <badref>), !dbg !DILocation(line: 234, scope: !483177)
2020-03-14T19:32:16.3631021Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3631650Z   %"[S.P.CoreLib]Internal.Runtime.EEType"* (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_NullableType
2020-03-14T19:32:16.3632234Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3633178Z     <badref> = call addrspace(0) %"[S.P.CoreLib]Internal.Runtime.EEType"* @S_P_CoreLib_Internal_Runtime_EEType__get_NullableType(i8* <badref>), !dbg !DILocation(line: 255, scope: !483177)
2020-03-14T19:32:16.3634146Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3634638Z   void (i8*, i8*)* @Object__GetRawData
2020-03-14T19:32:16.3635058Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3635779Z     call addrspace(0) void @Object__GetRawData(i8* <badref>, i8* %Temp5_), !dbg !DILocation(line: 265, scope: !483177)
2020-03-14T19:32:16.3636818Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3637561Z   void (i8*, i1)* @S_P_CoreLib_System_Diagnostics_Debug__Assert
2020-03-14T19:32:16.3638213Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3639079Z     call addrspace(0) void @S_P_CoreLib_System_Diagnostics_Debug__Assert(i8* <badref>, i1 <badref>), !dbg !DILocation(line: 255, scope: !483177)
2020-03-14T19:32:16.3639983Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3640898Z   void (i8*, i1)* @S_P_CoreLib_System_Diagnostics_Debug__Assert
2020-03-14T19:32:16.3641441Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3642368Z     call addrspace(0) void @S_P_CoreLib_System_Diagnostics_Debug__Assert(i8* <badref>, i1 %LdTemp1_), !dbg !DILocation(line: 252, scope: !483177)
2020-03-14T19:32:16.3643300Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3643881Z   void (i8*, i1)* @S_P_CoreLib_System_Diagnostics_Debug__Assert
2020-03-14T19:32:16.3644432Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3645273Z     call addrspace(0) void @S_P_CoreLib_System_Diagnostics_Debug__Assert(i8* <badref>, i1 <badref>), !dbg !DILocation(line: 248, scope: !483177)
2020-03-14T19:32:16.3646200Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3646783Z   void (i8*, i1)* @S_P_CoreLib_System_Diagnostics_Debug__Assert
2020-03-14T19:32:16.3647325Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3648170Z     call addrspace(0) void @S_P_CoreLib_System_Diagnostics_Debug__Assert(i8* <badref>, i1 %LdTemp0_), !dbg !DILocation(line: 234, scope: !483177)
2020-03-14T19:32:16.3649089Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3649788Z   i32 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_BaseSize
2020-03-14T19:32:16.3650305Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3651243Z     <badref> = call addrspace(0) i32 @S_P_CoreLib_Internal_Runtime_EEType__get_BaseSize(i8* <badref>), !dbg !DILocation(line: 264, scope: !483177)
2020-03-14T19:32:16.3652149Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3652753Z   i32 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_BaseSize
2020-03-14T19:32:16.3653286Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3654163Z     <badref> = call addrspace(0) i32 @S_P_CoreLib_Internal_Runtime_EEType__get_BaseSize(i8* <badref>), !dbg !DILocation(line: 237, scope: !483177)
2020-03-14T19:32:16.3655101Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3655689Z   i32 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_ValueTypeFieldPadding
2020-03-14T19:32:16.3656206Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3657057Z     <badref> = call addrspace(0) i32 @S_P_CoreLib_Internal_Runtime_EEType__get_ValueTypeFieldPadding(i8* <badref>), !dbg !DILocation(line: 264, scope: !483177)
2020-03-14T19:32:16.3658126Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3659674Z   i32 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_ValueTypeFieldPadding
2020-03-14T19:32:16.3660482Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3661381Z     <badref> = call addrspace(0) i32 @S_P_CoreLib_Internal_Runtime_EEType__get_ValueTypeFieldPadding(i8* <badref>), !dbg !DILocation(line: 237, scope: !483177)
2020-03-14T19:32:16.3662542Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3663099Z   void (i8*, i32, i64)* @RhpInitMultibyte
2020-03-14T19:32:16.3663588Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3664366Z     call addrspace(0) void @RhpInitMultibyte(i8* %Loadarg1_, i32 0, i64 %sub), !dbg !DILocation(line: 237, scope: !483177)
2020-03-14T19:32:16.3665234Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3666073Z   i1 (i8*, %"[S.P.CoreLib]Internal.Runtime.EEType"*, %"[S.P.CoreLib]Internal.Runtime.EEType"*)* @S_P_CoreLib_System_Runtime_RuntimeExports__UnboxAnyTypeCompare
2020-03-14T19:32:16.3666878Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3668149Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_System_Runtime_RuntimeExports__UnboxAnyTypeCompare(i8* <badref>, %"[S.P.CoreLib]Internal.Runtime.EEType"* %Loadloc0_23, %"[S.P.CoreLib]Internal.Runtime.EEType"* %Loadarg2_24), !dbg !DILocation(line: 252, scope: !483177)
2020-03-14T19:32:16.3669650Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3670374Z   i1 (i8*, %"[S.P.CoreLib]Internal.Runtime.EEType"*)* @S_P_CoreLib_Internal_Runtime_EEType__IsEquivalentTo
2020-03-14T19:32:16.3671002Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3672058Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__IsEquivalentTo(i8* <badref>, %"[S.P.CoreLib]Internal.Runtime.EEType"* %Loadloc0_42), !dbg !DILocation(line: 255, scope: !483177)
2020-03-14T19:32:16.3673273Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3673874Z   void (i8*, i8*)* @"S_P_CoreLib_Internal_Runtime_CompilerServices_Unsafe__As_0<UInt8__Bool>"
2020-03-14T19:32:16.3674456Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3675318Z     call addrspace(0) void @"S_P_CoreLib_Internal_Runtime_CompilerServices_Unsafe__As_0<UInt8__Bool>"(i8* <badref>, i8* %Temp3_), !dbg !DILocation(line: 258, scope: !483177)
2020-03-14T19:32:16.3676250Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3676809Z   i8 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_NullableValueOffset
2020-03-14T19:32:16.3677340Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3678161Z     <badref> = call addrspace(0) i8 @S_P_CoreLib_Internal_Runtime_EEType__get_NullableValueOffset(i8* <badref>), !dbg !DILocation(line: 261, scope: !483177)
2020-03-14T19:32:16.3679259Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3680021Z   void (i8*, i8*, i32)* @"S_P_CoreLib_Internal_Runtime_CompilerServices_Unsafe__Add<UInt8>"
2020-03-14T19:32:16.3680644Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3681719Z     call addrspace(0) void @"S_P_CoreLib_Internal_Runtime_CompilerServices_Unsafe__Add<UInt8>"(i8* <badref>, i8* %Temp4_, i32 %CastZIntget_NullableValueOffset_return), !dbg !DILocation(line: 261, scope: !483177)
2020-03-14T19:32:16.3682861Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3683464Z   i1 (i8*)* @S_P_CoreLib_Internal_Runtime_EEType__get_HasGCPointers
2020-03-14T19:32:16.3684023Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3684909Z     <badref> = call addrspace(0) i1 @S_P_CoreLib_Internal_Runtime_EEType__get_HasGCPointers(i8* <badref>), !dbg !DILocation(line: 267, scope: !483177)
2020-03-14T19:32:16.3685891Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3686465Z   void (i8*, i8*, i64)* @RhBulkMoveWithWriteBarrier
2020-03-14T19:32:16.3686973Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3687918Z     call addrspace(0) void @RhBulkMoveWithWriteBarrier(i8* %Loadarg1_80, i8* %Loadloc2_, i64 %Loadloc1_), !dbg !DILocation(line: 270, scope: !483177)
2020-03-14T19:32:16.3688913Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3689437Z   i8** @t_pShadowStackTop
2020-03-14T19:32:16.3689878Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3690654Z     store i8* %shadowStackTop82, i8** @t_pShadowStackTop, !dbg !DILocation(line: 270, scope: !483177)
2020-03-14T19:32:16.3691449Z   Global is referenced by parentless instruction!
2020-03-14T19:32:16.3691953Z   i8** @t_pShadowStackTop
2020-03-14T19:32:16.3692377Z   ; ModuleID = 'netscripten'
2020-03-14T19:32:16.3693100Z     store i8* %shadowStackTop, i8** @t_pShadowStackTop, !dbg !DILocation(line: 237, scope: !483177)
2020-03-14T19:32:16.3898428Z EXEC : LLVM error : Broken module found, compilation aborted! [F:\workspace\_work\1\s\tests\src\Simple\HelloWasm\HelloWasm.csproj]
2020-03-14T19:32:16.4489059Z F:\workspace\_work\1\s\bin\Windows_NT.x64.Debug\build\Microsoft.NETCore.Native.targets(230,5): error MSB3073: The command ""F:\workspace\_work\1\s\tests\\..\bin\Windows_NT.x64.Debug\tools\ilc" @"F:\workspace\_work\1\s\tests\src\Simple\HelloWasm\obj\Debug\x64\native\HelloWasm.ilc.rsp"" exited with code 1. [F:\workspace\_work\1\s\tests\src\Simple\HelloWasm\HelloWasm.csproj]
```

yowl commented:
@MichalStrehovsky Removing that code is fine here, the tests pass.  It should and has been done in local branches before, https://github.com/dotnet/corert/issues/6412#issuecomment-426133219, but no PR was ever created for it.  I do notice that `RhpInitMultibyte` is in InternalCalls.cs as
```
internal static extern unsafe void RhpInitMultibyte(ref byte dmem, int c, nuint size);
```
 but in GCMemoryHelpers.cpp its 
```
//         can be properly translated to a managed exception.
COOP_PINVOKE_CDECL_HELPER(void *, RhpInitMultibyte, (void * mem, int c, size_t size))
```
And I think that's why the linker complains:
```
wasm-ld : warning : function signature mismatch: RhpInitMultibyte [E:\GitHub\corert-managedunbox\tests\src\Simple\HelloWasm\HelloWasm.csproj]
  >>> defined as (i32, i32, i32) -> void in lto.tmp
  >>> defined as (i32, i32, i32) -> i32 in E:\GitHub\corert-managedunbox\tests\\..\bin\WebAssembly.wasm.Debug/sdk/libPortableRuntime.a(GCMemoryHelpers.cpp.o)
```

MichalStrehovsky commented:
> I do notice that RhpInitMultibyte is in InternalCalls.cs as

Ah. I copy pasted that signature from another place thinking that's fine, but that one was dead code. I fixed the sig and deleted the dead one. Thanks!
